// @composio/client/src/resources/link.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';

export class Link extends APIResource {
  /**
   * Creates a new authentication link session that users can use to connect their
   * accounts
   */
  create(body: LinkCreateParams, options?: RequestOptions): APIPromise<LinkCreateResponse> {
    return this._client.post('/api/v3/connected_accounts/link', { body, ...options });
  }
}

export interface LinkCreateResponse {
  /**
   * The connected account ID that was created
   */
  connected_account_id: string;

  /**
   * ISO timestamp when the link expires
   */
  expires_at: string;

  /**
   * The generated link token for the auth session
   */
  link_token: string;

  /**
   * The redirect URI to send users to for authentication
   */
  redirect_url: string;
}

export interface LinkCreateParams {
  /**
   * The auth config id to create a link for
   */
  auth_config_id: string;

  /**
   * The user id to create a link for
   */
  user_id: string;

  /**
   * The callback url to create a link for
   */
  callback_url?: string;
}

export declare namespace Link {
  export { type LinkCreateResponse as LinkCreateResponse, type LinkCreateParams as LinkCreateParams };
}
